apm-agent-plugins/apm-log-shader-plugin/apm-logback-plugin/apm-logback-plugin-impl/src/main/java/co/elastic/apm/agent/logback/LogbackLogShadingHelper.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.logback;
+
+import ch.qos.logback.classic.LoggerContext;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.FileAppender;
+import ch.qos.logback.core.rolling.FixedWindowRollingPolicy;
+import ch.qos.logback.core.rolling.RollingFileAppender;
+import ch.qos.logback.core.rolling.SizeBasedTriggeringPolicy;
+import co.elastic.apm.agent.log.shader.AbstractLogShadingHelper;
+import co.elastic.apm.agent.log.shader.Utils;
+import co.elastic.logging.logback.EcsEncoder;
+
+class LogbackLogShadingHelper extends AbstractLogShadingHelper<FileAppender<ILoggingEvent>> {
+
+    private static final LoggerContext defaultLoggerContext = new LoggerContext();
+
+    private static final LogbackLogShadingHelper INSTANCE = new LogbackLogShadingHelper();
+
+    static LogbackLogShadingHelper instance() {
+        return INSTANCE;
+    }
+
+    private LogbackLogShadingHelper() {}
+
+    @Override
+    protected String getAppenderName(FileAppender<ILoggingEvent> appender) {
+        return appender.getName();
+    }
+
+    @Override
+    protected FileAppender<ILoggingEvent> createAndConfigureAppender(FileAppender<ILoggingEvent> originalAppender, String appenderName) {
+        RollingFileAppender<ILoggingEvent> shadeAppender = new RollingFileAppender<>();
+        String shadeFile = Utils.computeShadeLogFilePath(originalAppender.getFile());
+        shadeAppender.setFile(shadeFile);
+
+        EcsEncoder ecsEncoder = new EcsEncoder();
+        ecsEncoder.setServiceName(getServiceName());
+        ecsEncoder.setIncludeMarkers(false);
+        ecsEncoder.setIncludeOrigin(false);
+        ecsEncoder.setStackTraceAsArray(false);
+        shadeAppender.setEncoder(ecsEncoder);
+
+        FixedWindowRollingPolicy rollingPolicy = new FixedWindowRollingPolicy();
+        rollingPolicy.setMinIndex(1);
+        rollingPolicy.setMaxIndex(1);
+        rollingPolicy.setFileNamePattern(shadeFile + ".%i");
+        rollingPolicy.setParent(shadeAppender);
+        rollingPolicy.setContext(defaultLoggerContext);
+        rollingPolicy.start();
+        shadeAppender.setRollingPolicy(rollingPolicy);
+
+        SizeBasedTriggeringPolicy<ILoggingEvent> triggeringPolicy = new SizeBasedTriggeringPolicy<>();
+        try {
+            VersionUtils.setMaxFileSize(triggeringPolicy, getMaxLogFileSize());
+        } catch (Throwable throwable) {
+            // We cannot log here because this plugin escapes slf4j package reallocation.
+            System.out.println("Failed to set max file size for log shader file-rolling strategy. Using the default " +
```

Comment:
why do we have to do the shading trick for slf4j on this module? could we just remove this portion of `apm-logback-plugin-impl/pom.xml`? It does not use any slf4j classes in a context where we need to reference the target cl's slf4j.

```xml
<relocation>
    <pattern>org.slf4j</pattern>
    <shadedPattern>co.elastic.apm.agent.logshading.shaded.slf4j</shadedPattern>
</relocation>
```

Reply:
Not sure now, maybe just a copy+paste+easier-to-just-leave 🙂 
I'll remove and see if anything breaks

Reply:
We cannot add an slf4j dependency in the logback module, but we will add a "facade" logging method in the abstract helper that the helper subclasses can use. The common module should be able to have this dependency